pyvrp/cpp/Solution.cpp

```diff
@@ -148,16 +148,31 @@ bool Solution::operator==(Solution const &other) const
 Solution::Solution(ProblemData const &data, RandomNumberGenerator &rng)
     : neighbours_(data.numLocations(), std::nullopt)
 {
+    std::vector<size_t> clients;
+
+    // First we add all required clients and some groupless optional clients...
+    for (size_t idx = data.numDepots(); idx != data.numLocations(); ++idx)
+    {
+        pyvrp::ProblemData::Client const &clientData = data.location(idx);
+
+        if (clientData.required)
+            clients.push_back(idx);
+        else if (!clientData.group && rng.randint(2))  // select randomly
+            clients.push_back(idx);
+    }
+
+    // ...but we select only one client per mutually exclusive group.
+    for (auto const &group : data.groups())
+        clients.push_back(group.clients()[rng.randint(group.size())]);
```

Comment:
We can also special-case this for optional groups, but I didn't really see the benefit there. But I don't know since I haven't really dealt with optional groups...